### PR TITLE
[WIP] Create subpackages for internal-ish code

### PR DESCRIPTION
**Pull request by steveluscher**
# Summary

The entire problem that this PR tries to solve is this.

1. Retain the simplicity of being able to `export * from '@solana/errors'` in `packages/library/`
2. Prevent things like `safeCaptureStackTrace()` from being exported from `packages/library/`

The steps taken to that end were: 

1. Change `moduleResolution` everywhere to `NodeNext`. This enables subpackage exports
2. Add `.js` extensions everywhere to satisfy that resolution regime
3. Create a subpackage export in `@solana/errors` for all of the stuff that we want to share internally but _not_ to export from `packages/library/`

Now any package in the `@solana` scope can use `@solana/errors/internal` but we can _still_ enjoy the simplicity of `export * from '@solana/errors'` in `packages/library/`

### CODE REVIEW

**Comment by steveluscher** [comment]
* **#2351** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2351?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @steveluscher and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/build-scripts/package.json [hunk 1]
```diff
@@ -3,7 +3,6 @@
     "version": "0.0.0",
     "private": true,
     "files": [
-        "add-js-extension-to-types.mjs",
```
Comment (steveluscher) [inline packages/build-scripts/package.json:6]: Nuked this, in favour of _actually_ adding the .js extension everywhere.

File: packages/build-scripts/getBaseConfig.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ export function getBaseConfig(platform: Platform, formats: Format[], _options: O
                               __REACTNATIVE__: `${platform === 'native'}`,
                               __VERSION__: `"${env.npm_package_version}"`,
                           },
-                          entry: [`./src/index.ts`],
+                          entry: ['./src/index.ts', './src/__internal.ts'],
```
Comment (steveluscher) [inline packages/build-scripts/getBaseConfig.ts:28]: This makes the build pick up this `__internal.ts` entrypoint if it exists.

File: packages/build-scripts/getBaseConfig.ts [hunk 3]
```diff
@@ -70,6 +70,7 @@ export function getBaseConfig(platform: Platform, formats: Format[], _options: O
                           platform: platform === 'node' ? 'node' : 'browser',
                           pure: ['process'],
                           sourcemap: format !== 'iife' || isDebugBuild,
+                          splitting: false,
```
Comment (steveluscher) [inline packages/build-scripts/getBaseConfig.ts:73]: This prevents `tsup` from doing fancy bundle splitting between entrypoints.

File: packages/errors/src/__internal.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+export * from './stack-trace.js';
```
Comment (steveluscher) [inline packages/errors/src/__internal.ts:1]: Here's that new ‘internal stuff’ entrypoint.

File: packages/rpc/src/rpc-integer-overflow-error.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
-import { safeCaptureStackTrace, SOLANA_ERROR__RPC__INTEGER_OVERFLOW, SolanaError } from '@solana/errors';
+import { SOLANA_ERROR__RPC__INTEGER_OVERFLOW, SolanaError } from '@solana/errors';
+import { safeCaptureStackTrace } from '@solana/errors/internal';
```
Comment (steveluscher) [inline packages/rpc/src/rpc-integer-overflow-error.ts:2]: Here's an example of making use of the internal subpackage.

File: packages/tsconfig/base.json [hunk 6]
```diff
@@ -9,7 +9,8 @@
         "forceConsistentCasingInFileNames": true,
         "inlineSources": false,
         "isolatedModules": true,
-        "moduleResolution": "node",
+        "module": "NodeNext",
+        "moduleResolution": "NodeNext",
```
Comment (steveluscher) [inline packages/tsconfig/base.json:13]: This is the fancy new resolution strategy that allows all of this to work.

File: packages/test-config/jest-unit.config.common.ts [hunk 7]
```diff
@@ -2,6 +2,11 @@ import { Config } from '@jest/types';
 import path from 'path';
 
 const config: Partial<Config.InitialProjectOptions> = {
+    moduleNameMapper: {
+        // Strip `.js` from imports
+        // https://github.com/swc-project/jest/issues/64#issuecomment-1029753225
+        '^(\\.{1,2}/.*)\\.js$': '$1',
+    },
```
Comment (steveluscher) [inline packages/test-config/jest-unit.config.common.ts:9]: Of course Jest doesn't know any of this, so we have to re-strip the .js extensions.

**Comment by lorisleiva** [comment]
Ugh it's so sad because I do love the end result but I cannot bear having to use `.js` extensions on `.ts` files. And the fact that we then have to strip them for Jest is the icing on the cake.

Is there no way we can do this as a build step instead?

**Comment by steveluscher** [comment]
> Is there no way we can do this as a build step instead?

Do what; add the extensions? No, because once you change to `"moduleResolution": "NodeNext"` it's gg.

**Comment by lorisleiva** [comment]
Well, adding `"moduleResolution": "NodeNext"` to our tsconfig changes how things are compiled, right?

What I'm saying is, instead of using the module resolution, can we not replicate the changes that this does at compile time to make subpackages work properly?

What exactly is `NodeNext` changing in our compiled code that makes subpackages possible?

**Comment by steveluscher** [comment]
> What exactly is NodeNext changing in our compiled code that makes subpackages possible?

I think subpackage imports just straight up don't work unless `moduleResolution` is `node16` or higher.

With `moduleResolution: NodeNext`:



With `moduleResolution: Node`:

**Comment by lorisleiva** [comment]
Oh I see, as a _consumer_ of our own API, we need `NodeNext` to import sub-packages. No matter how we export them. I'm happy to disagree and commit if everyone agrees that the benefits are worth it. `.js` extensions on `.ts` files do make me want to cry though.

**Comment by steveluscher** [comment]
Meh.

**Pull request closed without merging**